runtime/lua/vim/lsp/_inlay_hint.lua

```diff
@@ -149,26 +149,78 @@ end
 
 --- Clear inlay hints
 ---
----@param client_id integer|nil filter by client_id. All clients if nil
----@param bufnr integer|nil filter by buffer. All buffers if nil
+---@param bufnr (integer|nil) Buffer handle, or nil for current
 ---@private
-function M.clear(client_id, bufnr)
-  local buffers = bufnr and { resolve_bufnr(bufnr) } or vim.tbl_keys(hint_cache_by_buf)
-  for _, iter_bufnr in ipairs(buffers) do
-    M.__explicit_buffers[iter_bufnr] = false
-    local bufstate = hint_cache_by_buf[iter_bufnr]
-    local client_lens = (bufstate or {}).client_hint or {}
-    local client_ids = client_id and { client_id } or vim.tbl_keys(client_lens)
-    for _, iter_client_id in ipairs(client_ids) do
-      if bufstate then
-        bufstate.client_hint[iter_client_id] = {}
-      end
+local function clear(bufnr)
+  bufnr = resolve_bufnr(bufnr or 0)
+  local bufstate = hint_cache_by_buf[bufnr]
+  local client_lens = (bufstate or {}).client_hint or {}
+  local client_ids = vim.tbl_keys(client_lens)
+  for _, iter_client_id in ipairs(client_ids) do
+    if bufstate then
+      bufstate.client_hint[iter_client_id] = {}
     end
-    api.nvim_buf_clear_namespace(iter_bufnr, namespace, 0, -1)
   end
+  api.nvim_buf_clear_namespace(bufnr, namespace, 0, -1)
   vim.cmd('redraw!')
 end
 
+---@private
+local function reset_timer(reset_bufnr)
+  local timer = M.__buffers[reset_bufnr].timer
+  if timer then
+    M.__buffers[reset_bufnr].timer = nil
+    if not timer:is_closing() then
+      timer:stop()
+      timer:close()
+    end
+  end
+end
+
+---@private
+local function make_request(request_bufnr)
+  reset_timer(request_bufnr)
+  M.refresh({ bufnr = request_bufnr })
+end
+
+---@private
+function M.enable(bufnr)
+  bufnr = resolve_bufnr(bufnr or 0)
+  if not (M.__buffers[bufnr] and M.__buffers[bufnr].enabled) then
+    M.__buffers[bufnr] = { enabled = true, timer = nil }
+    M.refresh({ bufnr = bufnr })
+    api.nvim_buf_attach(bufnr, true, {
+      on_lines = function(_, cb_bufnr)
+        if not M.__buffers[cb_bufnr].enabled then
+          return true
+        end
+        reset_timer(cb_bufnr)
```

Comment:
not sure is the good way create timer for each buffer ,  and in `on_lines`  callback has `changedtick` param you can use  pass to request like `make_request(cb_bufnr, changedtick)`,  and when  lsp request callback running check `nvim_get_buf_chagnetick` make a compare. `if before_changetick < nvim_get_buf_changetick` do defer reqeust again.   some undo/redo also increase `changetick` and also some change without text chang also increase `changtick` .  this is my thought  and I do it on documentsymbol request . it combine 1000ms change to one request . 
 https://github.com/nvimdev/lspsaga.nvim/blob/3546dca5560b6868612acc567f6f3c07cf61c076/lua/lspsaga/symbol/init.lua#L25

Reply:
I don't like this either, is there a better way of remembering which buffers have inlay hint enabled (to handle refresh request)?

Reply:
That is what I was doing, except how do you cancel pending requests?

Reply:
i had point it out `changetick ` increase without text changed and this change will not trigger `on_lines` callback . and there has `on_changedtick` in `nvim_buf_attach` . these problem  i had comment in prev :-)